operator/src/main/java/oracle/kubernetes/operator/PodWatcher.java

```diff
@@ -199,107 +171,54 @@ public Step waitForDelete(V1Pod pod, Step next) {
-                  if (didResume.compareAndSet(false, true)) {
-                    fiber.resume(packet);
-                  }
-                };
-            register(metadata, ready);
-
-            // Timing window -- pod may have come ready before registration for callback
-            CallBuilderFactory factory =
-                ContainerResolver.getInstance().getContainer().getSpi(CallBuilderFactory.class);
-            fiber
-                .createChildFiber()
-                .start(
-                    factory
-                        .create()
-                        .readPodAsync(
-                            metadata.getName(),
-                            metadata.getNamespace(),
-                            new ResponseStep<V1Pod>(null) {
-                              @Override
-                              public NextAction onFailure(
-                                  Packet packet,
-                                  CallResponse<V1Pod> callResponse) {
-                                if (callResponse.getStatusCode() == CallBuilder.NOT_FOUND) {
-                                  return onSuccess(packet, callResponse);
-                                }
-                                return super.onFailure(packet, callResponse);
-                              }
-
-                              @Override
-                              public NextAction onSuccess(Packet packet, CallResponse<V1Pod> callResponse) {
-                                if (testPod(callResponse.getResult())) {
-                                  if (didResume.compareAndSet(false, true)) {
-                                    unregister(metadata, ready);
-                                    fiber.resume(packet);
-                                  }
-                                }
-                                return doNext(packet);
-                              }
-                            }),
-                    packet.clone(),
-                    null);
-          });
+    V1ObjectMeta getMetadata(V1Pod pod) {
+      return pod.getMetadata();
     }
-
-    protected void log(V1ObjectMeta metadata) {
-      // no-op
+    
+    @Override
+    Step createReadAsyncStep(String name, String namespace, ResponseStep<V1Pod> responseStep) {
+      return new CallBuilder().readPodAsync(name, namespace, responseStep);
     }
-
-    protected abstract boolean testPod(V1Pod result);
-
-    protected abstract void register(V1ObjectMeta metadata, Runnable callback);
-
-    protected abstract void unregister(V1ObjectMeta metadata, Runnable callback);
   }
 
   private class WaitForPodReadyStep extends WaitForPodStatusStep {
+
     private WaitForPodReadyStep(V1Pod pod, Step next) {
       super(pod, next);
     }
 
+    // A pod is ready if it is not being deleted and has the ready status.
+    @Override
+    protected boolean isReady(V1Pod result) {
+      return result != null && !PodHelper.isDeleting(result) && PodHelper.isReady(result);
```

Comment:
curious why you have to check for not deleting.

Reply:
That's unchanged from the old code. isDeleting returns true if there is a deletionTimestamp defined on the pod's metadata, which indicates that Kubernetes is shutting it down. Such a pod is never deemed "ready," even if the other conditions hold.

Reply:
This catches the case where we are deleting the current pod and then recreating a new pod with the same name for which we then want to wait for ready.  If the code to wait is too fast, then we would detect as "still" ready the pod that is being deleted.  This code catches that case.